File: src/audio/feedback.py
"""
Audio feedback tones for Rex voice assistant.

Plays short musical tones to indicate listening state:
- Ascending C→G: Rex is now listening
- Descending G→C: Rex has finished listening
- Looping D→A: Rex is thinking (waiting for LLM)

All tones are played through AudioManager for coordinated audio output.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np

if TYPE_CHECKING:
    from audio.manager import AudioManager

# Note frequencies (Hz)
C4 = 261.63  # Middle C
G4 = 392.00  # Perfect fifth above C4
D4 = 293.66  # D above middle C
A4 = 440.00  # A above middle C

# Tone parameters
SAMPLE_RATE = 44100
NOTE_DURATION = 0.1  # seconds per note
GAP_DURATION = 0.05  # seconds between notes

# Thinking tone parameters (slower tempo, softer)
THINKING_NOTE_DURATION = 0.4
THINKING_GAP_DURATION = 0.05
THINKING_VOLUME = 0.2


def _generate_tone(
    frequency: float,
    duration: float,
    sample_rate: int = SAMPLE_RATE,
    volume: float = 0.3,
    envelope_duration: float = 0.02,
) -> np.ndarray:
    """Generate a sine wave tone with smooth attack and release.

    Args:
        frequency: Tone frequency in Hz
        duration: Tone duration in seconds
        sample_rate: Audio sample rate
        volume: Volume multiplier (0.0 to 1.0)
        envelope_duration: Duration of fade-in/fade-out in seconds (default 20ms)
    """
    t = np.linspace(0, duration, int(sample_rate * duration), endpoint=False)
    tone = np.sin(2 * np.pi * frequency * t)

    # Apply raised cosine envelope to avoid clicks
    envelope_samples = int(envelope_duration * sample_rate)
    envelope = np.ones_like(tone)
    # Fade in: 0.5 * (1 - cos(pi * t)) goes from 0 to 1 smoothly
    envelope[:envelope_samples] = 0.5 * (1 - np.cos(np.pi * np.linspace(0, 1, envelope_samples)))
    # Fade out: 0.5 * (1 + cos(pi * t)) goes from 1 to 0 smoothly
    envelope[-envelope_samples:] = 0.5 * (1 + np.cos(np.pi * np.linspace(0, 1, envelope_samples)))

    return (tone * envelope * volume).astype(np.float32)


def _generate_two_tone_sequence(freq1: float, freq2: float) -> np.ndarray:
    """Generate a two-note sequence with a small gap between notes."""
    tone1 = _generate_tone(freq1, NOTE_DURATION)
    gap = np.zeros(int(SAMPLE_RATE * GAP_DURATION), dtype=np.float32)
    tone2 = _generate_tone(freq2, NOTE_DURATION)

    return np.concatenate([tone1, gap, tone2])


def generate_listening_tone() -> np.ndarray:
    """Generate ascending C→G tone to indicate Rex is listening."""
    return _generate_two_tone_sequence(C4, G4)


def generate_done_tone() -> np.ndarray:
    """Generate descending G→C tone to indicate Rex has finished listening."""
    return _generate_two_tone_sequence(G4, C4)


def generate_thinking_sequence() -> np.ndarray:
    """Generate a D→A tone sequence with slower timing for thinking feedback."""
    # Use longer 50ms envelope for smoother fade-in (reduces pop at loop start)
    tone1 = _generate_tone(D4, THINKING_NOTE_DURATION, volume=THINKING_VOLUME, envelope_duration=0.05)
    gap = np.zeros(int(SAMPLE_RATE * THINKING_GAP_DURATION), dtype=np.float32)
    tone2 = _generate_tone(A4, THINKING_NOTE_DURATION, volume=THINKING_VOLUME, envelope_duration=0.05)
    trailing_gap = np.zeros(int(SAMPLE_RATE * THINKING_GAP_DURATION), dtype=np.float32)
    return np.concatenate([tone1, gap, tone2, trailing_gap])


class ThinkingTone:
    """
    Looping D→A tone that plays while waiting for LLM inference.

    Can be used as a context manager:
        with ThinkingTone(audio_manager):
            # tone plays during this block
            result = slow_operation()

    Uses AudioManager's persistent stream for pop-free looping.
    """

    def __init__(self, audio_manager: AudioManager):
        self._audio_manager = audio_manager

    def __enter__(self):
        self.start()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.stop()
        return False  # Don't suppress exceptions

    def start(self):
        """Start playing the thinking tone. Non-blocking."""
        audio = generate_thinking_sequence()
        self._audio_manager.start_loop(audio)

    def stop(self):
        """Stop the thinking tone."""
        self._audio_manager.stop_loop()

File: internal/audio/feedback.go
package audio

import "math"

// Note frequencies in Hz.
const (
	C4 = 261.63 // Middle C
	G4 = 392.00 // Perfect fifth above C4
	D4 = 293.66 // D above middle C
	A4 = 440.00 // A above middle C
)

// Tone timing parameters.
const (
	noteDuration = 0.1  // seconds per note
	gapDuration  = 0.05 // seconds between notes

	thinkingNoteDuration = 0.4
	thinkingGapDuration  = 0.05
	thinkingVolume       = 0.2

	defaultVolume          = 0.3
	defaultEnvelopeDuration = 0.02 // 20ms fade in/out
	thinkingEnvelopeDuration = 0.05 // 50ms for smoother loop transitions
)

// generateTone creates a sine wave tone with smooth raised-cosine envelope.
func generateTone(frequency, duration float64, sampleRate int, volume, envelopeDuration float64) []float32 {
	numSamples := int(float64(sampleRate) * duration)
	if numSamples == 0 {
		return nil
	}

	samples := make([]float32, numSamples)
	envelopeSamples := int(envelopeDuration * float64(sampleRate))

	for i := 0; i < numSamples; i++ {
		t := float64(i) / float64(sampleRate)
		sine := math.Sin(2 * math.Pi * frequency * t)

		// Raised cosine envelope for click-free transitions.
		var env float64
		switch {
		case i < envelopeSamples:
			// Fade in: 0.5 * (1 - cos(pi * progress))
			progress := float64(i) / float64(envelopeSamples)
			env = 0.5 * (1 - math.Cos(math.Pi*progress))
		case i >= numSamples-envelopeSamples:
			// Fade out: 0.5 * (1 + cos(pi * progress))
			progress := float64(i-(numSamples-envelopeSamples)) / float64(envelopeSamples)
			env = 0.5 * (1 + math.Cos(math.Pi*progress))
		default:
			env = 1.0
		}

		samples[i] = float32(sine * env * volume)
	}

	return samples
}

// generateTwoToneSequence creates a two-note sequence with a silence gap between them.
func generateTwoToneSequence(freq1, freq2 float64, sampleRate int, noteDur, gapDur, volume, envDur float64) []float32 {
	tone1 := generateTone(freq1, noteDur, sampleRate, volume, envDur)
	gapSamples := int(float64(sampleRate) * gapDur)
	gap := make([]float32, gapSamples)
	tone2 := generateTone(freq2, noteDur, sampleRate, volume, envDur)

	result := make([]float32, 0, len(tone1)+len(gap)+len(tone2))
	result = append(result, tone1...)
	result = append(result, gap...)
	result = append(result, tone2...)
	return result
}

// GenerateListeningTone creates an ascending C->G tone indicating Rex is listening.
func GenerateListeningTone(sampleRate int) []float32 {
	return generateTwoToneSequence(C4, G4, sampleRate, noteDuration, gapDuration, defaultVolume, defaultEnvelopeDuration)
}

// GenerateDoneTone creates a descending G->C tone indicating Rex finished listening.
func GenerateDoneTone(sampleRate int) []float32 {
	return generateTwoToneSequence(G4, C4, sampleRate, noteDuration, gapDuration, defaultVolume, defaultEnvelopeDuration)
}

// GenerateThinkingSequence creates a D->A tone sequence for looping during LLM inference.
// Uses slower timing and softer volume than the listening/done tones, with a trailing
// gap for seamless looping.
func GenerateThinkingSequence(sampleRate int) []float32 {
	seq := generateTwoToneSequence(D4, A4, sampleRate, thinkingNoteDuration, thinkingGapDuration, thinkingVolume, thinkingEnvelopeDuration)
	trailingGap := make([]float32, int(float64(sampleRate)*thinkingGapDuration))
	return append(seq, trailingGap...)
}
